path_dynamic_roles.go

```diff
@@ -0,0 +1,313 @@
+					Callback:                    b.pathDynamicRoleDelete,
+					ForwardPerformanceStandby:   true,
+					ForwardPerformanceSecondary: true,
+				},
+			},
+			HelpSynopsis:    staticRoleHelpSynopsis,
+			HelpDescription: staticRoleHelpDescription,
+		},
+		// LIST role
+		{
+			Pattern: dynamicRolePath + "?$",
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ListOperation: &framework.PathOperation{
+					Callback: b.pathDynamicRoleList,
+				},
+			},
+			HelpSynopsis:    "List all the dynamic roles Vault is currently managing in OpenLDAP.",
+			HelpDescription: "List all the dynamic roles being managed by Vault.",
+		},
+		// GET credentials
+		{
+			Pattern: path.Join(dynamicCredPath, framework.MatchAllRegex("name")),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "Name of the dynamic role.",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.pathDynamicCredsRead,
+				},
+			},
+			HelpSynopsis: "Request LDAP credentials for a dynamic role. These credentials are " +
+				"created within OpenLDAP when querying this endpoint.",
+			HelpDescription: "This path requests new LDAP credentials for a certain dynamic role. " +
+				"The credentials are created within OpenLDAP based on the creation_ldif specified " +
+				"within the dynamic role configuration.",
+		},
+	}
+}
+
+func dynamicSecretCreds(b *backend) *framework.Secret {
+	return &framework.Secret{
+		Type: secretCredsType,
+		Fields: map[string]*framework.FieldSchema{
+			"username": {
+				Type:        framework.TypeString,
+				Description: "Username of the generated account",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Password to access the generated account",
+			},
+			"distinguished_names": {
+				Type: framework.TypeStringSlice,
+				Description: "List of the distinguished names (DN) created. Each name in this list corresponds to" +
+					"each action taken within the creation_ldif statements. This does not de-duplicate entries, " +
+					"so this will have one entry for each LDIF statement within creation_ldif.",
+			},
+		},
+
+		Renew:  b.secretCredsRenew(),
+		Revoke: b.secretCredsRevoke(),
+	}
+}
+
+func (b *backend) pathDynamicRoleCreateUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	rawData := data.Raw
+	err := convertToDuration(rawData, "default_ttl", "max_ttl")
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert TTLs to duration: %w", err)
+	}
+
+	dRole := &dynamicRole{}
+	err = mapstructure.WeakDecode(rawData, dRole)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode request: %w", err)
+	}
+
+	err = validateDynamicRole(dRole)
+	if err != nil {
+		return nil, err
+	}
+
+	dRole.CreationLDIF = decodeBase64(dRole.CreationLDIF)
```

Comment:
Does this mean that the LDIFs will always be provided as base64 encoded values? Your unit and acceptance tests seem to be able to provide them as non-encoded strings, but I'm a bit puzzled by how this is working if `base64.StdEncoding.DecodeString` is always being called on a non-empty string.

Reply:
No, they are optionally base64 encoded.

This works because `base64.StdEncoding.DecodeString` will error if the string in question is not base64 encoded. A plain LDIF string cannot be base64 decoded so it will fail. That error will result in `decodeBase64` returning the original string. It's pretty much the equivalent of "if string is base64 encoded, decode it and return, otherwise return the original" except the conditional and the decoding is happening at the same step.

I'll add a comment on the `decodeBase64` function to try to make this clearer in the code.

Reply:
Oh I didn't realize that we were ignoring the error and returning `str` if so.

Reply:
I know that we call `assertValidLDIFTemplate` down below, but I wonder if we should `len` check `dRole.CreationLDIF` early and return and better error here indicating that the creation LDIF is required.

Reply:
🤦 It's handled under `validateDynamicRole`

Reply:
Looking at this again, I think I'm going to shuffle this around a little bit:
1. Move the base64 decoding above the validation
2. Move the LDIF template assertion into `validateDynamicRole` so all of the validation logic is in one function